controllers/mdutil/util.go

```diff
@@ -472,6 +472,21 @@ func GetActualReplicaCountForMachineSets(machineSets []*clusterv1.MachineSet) in
 	return totalActualReplicas
 }
 
+// GetTotalReplicaCountForMachineSets returns sum of max(ms.Spec.Replicas, ms.Status.Replicas) across all the machine sets.
+// This is used to guarantee that the total number of machines will not exceed md.Spec.Replicas + maxSurge
+// Use max(spec.Replicas,status.Replicas) to cover the cases that:
+// 1. Scale up, where spec.Replicas increased but no machine created yet, so spec.Replicas > status.Replicas
+// 2. Scale down, where spec.Replicas decreased but machine not deleted yet, so spec.Replicas < status.Replicas
+func GetTotalReplicaCountForMachineSets(machineSets []*clusterv1.MachineSet) int32 {
+	totalReplicas := int32(0)
+	for _, ms := range machineSets {
+		if ms != nil {
+			totalReplicas += max(*(ms.Spec.Replicas), ms.Status.Replicas)
```

Comment:
Can we use the min/max functions exposed in https://github.com/kubernetes/utils/blob/3bc9d57fc79286ce09cc68ac72e5345394af9d96/integer/integer.go?